packages/cw721/src/msg.rs

```diff
@@ -21,15 +60,931 @@ pub enum Cw721ExecuteMsg {
+        include_expired: Option<bool>,
+        start_after: Option<String>,
+        limit: Option<u32>,
+    },
+    /// Total number of tokens issued
+    #[returns(NumTokensResponse)]
+    NumTokens {},
+
+    #[deprecated(
+        since = "0.19.0",
+        note = "Please use GetCollectionInfoAndExtension instead"
+    )]
+    #[returns(CollectionInfoAndExtensionResponse<TCollectionExtension>)]
+    /// Deprecated: use GetCollectionInfoAndExtension instead! Will be removed in next release!
+    ContractInfo {},
+
+    /// Returns `CollectionInfoAndExtensionResponse`
+    #[returns(CollectionInfoAndExtensionResponse<TCollectionExtension>)]
+    GetCollectionInfoAndExtension {},
+
+    /// returns `AllInfoResponse` which contains contract, collection and nft details
+    #[returns(AllInfoResponse)]
+    GetAllInfo {},
+
+    /// Returns `CollectionExtensionAttributes`
+    #[returns(CollectionExtensionAttributes)]
+    GetCollectionExtensionAttributes {},
+
+    #[deprecated(since = "0.19.0", note = "Please use GetMinterOwnership instead")]
+    #[returns(Ownership<Addr>)]
+    /// Deprecated: use GetMinterOwnership instead! Will be removed in next release!
+    Ownership {},
+
+    /// Return the minter
+    #[deprecated(since = "0.19.0", note = "Please use GetMinterOwnership instead")]
+    #[returns(MinterResponse)]
+    /// Deprecated: use GetMinterOwnership instead! Will be removed in next release!
+    Minter {},
+
+    #[returns(Ownership<Addr>)]
+    GetMinterOwnership {},
+
+    #[returns(Ownership<Addr>)]
+    GetCreatorOwnership {},
+
+    /// With MetaData Extension.
+    /// Returns metadata about one particular token, based on *ERC721 Metadata JSON Schema*
+    /// but directly from the contract
+    #[returns(NftInfoResponse<TNftExtension>)]
+    NftInfo { token_id: String },
+
+    #[returns(Option<NftInfoResponse<TNftExtension>>)]
+    GetNftByExtension { extension: TNftExtension },
+
+    /// With MetaData Extension.
+    /// Returns the result of both `NftInfo` and `OwnerOf` as one query as an optimization
+    /// for clients
+    #[returns(AllNftInfoResponse<TNftExtension>)]
+    AllNftInfo {
+        token_id: String,
+        /// unset or false will filter out expired approvals, you must set to true to see them
+        include_expired: Option<bool>,
+    },
+
+    /// With Enumerable extension.
+    /// Returns all tokens owned by the given address, [] if unset.
+    #[returns(TokensResponse)]
+    Tokens {
+        owner: String,
+        start_after: Option<String>,
+        limit: Option<u32>,
+    },
+    /// With Enumerable extension.
+    /// Requires pagination. Lists all token_ids controlled by the contract.
+    #[returns(TokensResponse)]
+    AllTokens {
+        start_after: Option<String>,
+        limit: Option<u32>,
+    },
+
+    /// Custom msg query. Default implementation returns an empty binary.
+    #[returns(())]
+    Extension { msg: TExtensionQueryMsg },
```

Comment:
Fixes previous design flaw where generic was used in `Extension`, `NftInfo`, and `AllNftInfo` query msgs:

```rs
...
    #[returns(cw721::NftInfoResponse<Q>)]
    NftInfo { token_id: String },
...
    #[returns(cw721::AllNftInfoResponse<Q>)]
    AllNftInfo {
        token_id: String,
        /// unset or false will filter out expired approvals, you must set to true to see them
        include_expired: Option<bool>,
    },
...
    #[returns(())]
    Extension { msg: Q },
...
```

There are now `TNftExtension` and `TExtensionQueryMsg`:

```rs
...
    #[returns(NftInfoResponse<TNftExtension>)]
    NftInfo { token_id: String },
...
    #[returns(Option<NftInfoResponse<TNftExtension>>)]
    GetNftByExtension { extension: TNftExtension },
...
    #[returns(AllNftInfoResponse<TNftExtension>)]
    AllNftInfo {
        token_id: String,
        /// unset or false will filter out expired approvals, you must set to true to see them
        include_expired: Option<bool>,
    },
...
    #[returns(())]
    Extension { msg: TExtensionQueryMsg },
...
```

Please also note, there's a new `GetNftByExtension` which uses, the only msg using `TNftExtension` as an argument. Otherwise compiler cannot infer `TNftExtension` - since `NftInfo` and `AllNftInfo` msg has only `TNftExtension` as return value.

@JakeHartnell

Reply:
Noted above, but you be sure to document what `GetNftByExtension` is intended to be used for?